man/cidupdate.1
.\" %W% %G%
.TH CIDUPDATE 1
.SH NAME
.B cidupdate
- update aliases in the NCID call file
.SH SYNOPSIS
.B cidupdate
[-A aliasfile] [-C cidlog] [newcidlog]
.SH DESCRIPTION
The
.B cidupdate
script updates the cidcall.log file using the aliases found in the
ncidd.alias file.
.SH "OPTIONS"
.PD 0
.TP 14
.B -A aliasfile
Set the alias file to \fIaliasfile\fR
.br
Default: /etc/ncid/ncidd.alias
.TP
.B -C cidlog
Set the call file to \fIcidlog\fR
.br
Default: /var/log/cidcall.log
.PD
.SS "Arguments"
.PD 0
.TP 14
.B newcidlog
The updated call file.
.br
Default: <cidlog>.new
.PD
.SH SEE ALSO
ncidd.conf.5,
ncidd.alias.5,
ncidd.blacklist.5,
cidcall.1,
cidupdate.1


doc/CHANGES
ncid - Network CallerID

Version 0.85
    - README: updated
    - Makefile: updated
    - renamed cidgate/ to gateway/
    - client/Makefile: updated
    - client/ncid.sh:
        added verbose levels 1-9
        added doVerbose{}
        changed all "if $Verbose" to "doVerbose <string> 1"
        added test for $ProgDir == "" to correctly set $Program from $ProgName
        updated comments for $ProgDir and $ProgName
        added --alt-date|-A option to display date as dd/mm/yyyy
        added country code for HR (Croatia)
        fixed "call" array pronlem that crashed ncid if call
            not found in the array.
        added date separator ($DateSepar) so "-" or "." can be used
            in place of "/" in displayed date
        changed --message option to do nothing, messages always sent
            to the output module
        removed -M | --messages option from Usage:
        added code to set ring indirectly from a module
            name with "-" replaced by "_" in ncid.conf
        getCID{}: removed verbose inside "if $Raw {"
                  added line: if {$Program != ""} {doVerbose "$dataBlock" 1}
    - client/ncid.conf-in:
            added AltDate configuration variable
            added ring number configuration
    - debian/README.Debian: updated
    - debian/Makefile: updated
    - debian/ncid-*.init: updated
    - debian/ncid-notify.init: new
    - debian/postrm-client: updated
    - debian/prerm-client: updated
    - debian/rules: updated
    - doc/INSTALL-andLinux: removed
    - doc/INSTALL-Cygwin: added rebase and service documentation
    - doc/INSTALL-Fedora: updated
    - doc/INSTALL-Mac: added section to install using launchctl
    - doc/INSTALL-TiVo: updated for the startncid script.
    - doc/INSTALL-Ubuntu: added section to install from GetDeb
    - doc/INSTALL-Win: updated
    - doc/Verbose-ncidd: updated
    - doc/NCID-SDK.odt: updated
    - Fedora/Makefile: updated
    - Fedora/ncid.spec: updated
    - Fedora/ncid-notify.init: new
    - Fedora/ncid-notify.service-in: new
    - Fedora/ncidd.service-in: changed network.target to multi-user.target
    - Fedora/ncid2ncid.service-in: changed network.target to multi-user.target
    - Fedora/ncid2sip.service-in: changed network.target to multi-user.target
    - Fedora/ncid2yac.service-in: changed network.target to multi-user.target
    - FreeBSD/Makefile: updated
    - FreeBSD/ncid-notify.rc: new
    - gateway/Makefile: updated
    - gateway/sip2ncid.c:
        added print of command line options to logfile
        doPCAP(): - changed  errorExit() call
        getOptions(): - added "--osx-launchd" for Mac launchd
        parseLines: fixed linenum[0] never populated
    - gateway/sip2ncid.h: updated Usage: 
    - gateway/ncid2ncid.c:
        added "OUT: " & "HUP: " to lineType[]
       added print of command line options to logfile
       processPackets(): moved if (outcall) to inside above if statement
       getOptions(): - added "--osx-launchd" for Mac launchd
    - gateway/ncid2ncid.h: updated Usage: 
    - Mac/*: new
    - man/Makefile: updated
    - man/ncid.1: updated
    - man/notify.1: new
    - man/ncid.conf.5: updated
    - man/ncid2ncid.1: updated
    - man/ncidd.8: updated
    - man/ncidd.conf.5: Updated
    - man/ncidd.blacklist.5: updated
    - man/ncidd.whitelist.5: new
    - man/ncidmodules.1: removed
    - man/ncidmodules.conf.5: removed
    - man/ncid-modules.7: new
    - man/ncid-tivo.1: moved to TiVo/
    - man/ncid-*.1: rewrote
    - mancid*.1: rewrote
    - man/ncidrotate.1: moved to scripts/
    - man/ncidtoold.1: removed
    - man/sip2ncid.conf.5: Updated
    - modules/Makefile.modules: updated
    - modules/README.modules: updated
    - modules/ncid-initmodem.sh: modified comments
    - modules/ncid-kpopup.sh: modified comments
    - modules/ncid-mythtv.sh: modified comments
    - modules/ncid-notify.sh: new
    - modules/ncid-tivo.sh: moved to TiVo/
    - modules/*.sh: lots of code minor changes in all modules
    - modules/ncidmodules.conf-in: removed
    - modules/ncid-*.conf: new
    - scripts/README.scripts: updated
    - scripts/README.scripts: renamed to README.logrotate
    - scripts/Makefile: updated
    - scripts/ncidrotate.1: moved from man/
    - scripts/ncid.logrotate: updated to work with all service files & systemctl
    - scripts/ncidrotate.sh: improved code a bit
    - renamed scripts/ to logrotate/
    - server/Makefile: updated
    - server/README.server: updated
    - server/ncidd.blacklist: added 4 area codes to block
    - server/ncidd.whitelist: new
    - server/ncidd.c:
        changed ringline to infoline, initialize to ONELINE instead of "-"
        Updated CIDINFO line comment for sendInfo(int mainsock)
        main() - added: strncpy(infoline, lineid, CIDSIZE - 1);
        removed extern void builtinAlias(), userAlias();
        Moved position of command line printout in logfile
        formatCID() - changed callout to calltype with values IN, OUT, and HUP
                    - removed builtinAlias() & userAlias()
                    - added test for "CALLHUP" in the CALL line from a gateway
                    - added switch statement for linelabel value
        main(): - added call whitelist lines under if (hangup)
        getOptions(): - added whitelist option
                      - added "--osx-launchd" for Mac launchd
        reload(): - moved blacklist reload inside if (hangup)
                  - added whitelist reload
        main(): - added code to print total alias entries
    - server/ncidd.conf-in: added whitelist variable
    - server/ncidd.h: - added defines for IN, OUT, and HUP
                      - added whitelist option to USAGE define
                      - replaced HANGUPMSG define with BLMSG and WLMSG
                      - updated Usage:
    - server/nciddalias.c:idoAlias(): - added Alias Entries log message
                           - added builtinAlias() & userAlias() was in ncidd.c 

tools/cidupdate.pl
#!/usr/bin/perl

# cidupdate - update Caller ID aliases in alias file

# Created by Aron Green on Mon Nov 25, 2002
# based on cidlog and cidalias
# Cleanup by John L. Chmielewski on Tue Nov 26, 2002
# Changed $CONFIG value on Sat May 21, 2005 by John L. Chmielewski
# Modified by John L. Chmielewski on Sat Aug 13, 2005
#   - Changed from using config file to alias file
# Modified by John L. Chmielewski on Fri Apr 14, 2006
#   - Changed name from cidlogupd to cidupdate, updated variable names
# Modified by John L. Chmielewski on Sat Aug 12, 2006
#   - Fixed "-c cidlog" option, Fixed alias expression line to handle
#     '*' for a name or number, fixed alias checking in if statements
# Modified by John L. Chmielewski on Sun Feb 13, 2011
#   - changed the -a option to -A and the -c option to -C
#   - removed unused -l option
# Modified by John L. Chmielewski on Sat Jun 11, 2011
#   - fixed regex and changed join and split character from ':' to '"'
#   - changed regex for getting the number to include special characters

use Getopt::Std;

$ALIAS = "/etc/ncid/ncidd.alias";
$CIDLOG = "/var/log/cidcall.log";

getopts('A:C:') ||
    die "Usage: cidupdate [-A aliasfile] [-C cidlog] [newcidlog]\n";

($alias = $opt_A) || ($alias = $ALIAS);
($cidlog = $opt_C) || ($cidlog = $CIDLOG);
($newcidlog = shift) || ($newcidlog = sprintf("%s.new", $cidlog));

open(ALIASFILE, $alias) || die "Could not open $alias\n";
open(CIDLOG, $cidlog) || die "Could not open $cidlog\n";
open(NEWCIDLOG, ">$newcidlog") || die "Could not open $newcidlog\n";

while (<ALIASFILE>) {
    if (/^alias/) {
        chomp;
        ($type, $from, $to, $value) = /^.*alias\s+(\w+)\s+"*([^"]+)"*\s+=\s+"*([^"]+)"*\s+if\s+"*([^"]+)"*$/;
        if ($value == "") {
        ($type, $from, $to) = /^.*alias\s+(\w+)\s+"*([^"]+)"*\s+=\s+"*([^"]+)"*(.*)$/;
        }
        $alias = join('"', ($type, $from, $to, $value));
        push(@aliases, $alias);
    }
}

#CID: *DATE*11242002*TIME*2112*LINE*1*NMBR*9549142285*MESG*NONE*NAME*Cell*

while (<CIDLOG>) {
    if (/CID|EXTRA/) {
        ($date, $time, $number, $mesg, $name) = 
            /.*DATE.(\d+).*TIME.(\d+).*NU*MBE*R.(.*)\*MESG.(\w+).*NAME.(.*)\*+$/;
        ($line) = /.*LINE.([-\w\s]+).*/;

    foreach $alias (@aliases) {
        ($type, $from, $to, $value) = split(/"/, $alias);
        if ($value ne "") {
            if ($type eq "NAME" && $number eq $value) {$name = $to;}
            if ($type eq "NMBR" && $name eq $value) {$number = $to;}
        } else {
            if ($type eq "NAME" && $name eq $from) {$name = $to;}
            if ($type eq "NMBR" && $number eq $from) {$number = $to;}
        }
    }
    printf(NEWCIDLOG
        "CID: *DATE*%s*TIME*%s*LINE*%s*NMBR*%s*MESG*%s*NAME*%s*\n",
        $date, $time, $line, $number, $mesg, $name);
    } else {
        printf(NEWCIDLOG);
    }
}
print "diff $cidlog $newcidlog\n";
exec('diff', $cidlog, $newcidlog);
